# Árvores II - Tarefa 2

### 1. Carregar as bases

Vamos carregar as bases lidas na tarefa passada. Se você salvou essas bases em arquivo texto, basta fazer a leitura com o comando ```pd.read_csv``` das seguintes bases:

- X_train
- Y_train
- X_test
- Y_test

Não se esqueça de considerar a leitura dos índices dos arquivos no ```read_csv()```!

In [56]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In [47]:
# Carregar os dados de treino e teste
caminho_treino_x = 'X_train.csv'
dados_treino_x  = pd.read_csv(caminho_treino_x)
dados_treino_x.head()

('tBodyAcc-mean()-X',)  ('tBodyAcc-mean()-Y',)  ('tBodyAcc-mean()-Z',)  \
0                0.288585               -0.020294               -0.132905   
1                0.278419               -0.016411               -0.123520   
2                0.279653               -0.019467               -0.113462   
3                0.279174               -0.026201               -0.123283   
4                0.276629               -0.016570               -0.115362   

   ('tBodyAcc-std()-X',)  ('tBodyAcc-std()-Y',)  ('tBodyAcc-std()-Z',)  \
0              -0.995279              -0.983111              -0.913526   
1              -0.998245              -0.975300              -0.960322   
2              -0.995380              -0.967187              -0.978944   
3              -0.996091              -0.983403              -0.990675   
4              -0.998139              -0.980817              -0.990482   

   ('tBodyAcc-mad()-X',)  ('tBodyAcc-mad()-Y',)  ('tBodyAcc-mad()-Z',)  \
0              -0.995112              -0.983185              -0.923527   
1              -0.998807              -0.974914              -0.957686   
2              -0.996520              -0.963668              -0.977469   
3              -0.997099              -0.982750              -0.989302   
4              -0.998321              -0.979672              -0.990441   

   ('tBodyAcc-max()-X',)  ...  ('fBodyBodyGyroJerkMag-meanFreq()',)  \
0              -0.934724  ...                             -0.074323   
1              -0.943068  ...                              0.158075   
2              -0.938692  ...                              0.414503   
3              -0.938692  ...                              0.404573   
4              -0.942469  ...                              0.087753   

   ('fBodyBodyGyroJerkMag-skewness()',)  ('fBodyBodyGyroJerkMag-kurtosis()',)  \
0                             -0.298676                             -0.710304   
1                             -0.595051                             -0.861499   
2                             -0.390748                             -0.760104   
3                             -0.117290                             -0.482845   
4                             -0.351471                             -0.699205   

   ('angle(tBodyAccMean,gravity)',)  \
0                         -0.112754   
1                          0.053477   
2                         -0.118559   
3                         -0.036788   
4                          0.123320   

   ('angle(tBodyAccJerkMean),gravityMean)',)  \
0                                   0.030400   
1                                  -0.007435   
2                                   0.177899   
3                                  -0.012892   
4                                   0.122542   

   ('angle(tBodyGyroMean,gravityMean)',)  \
0                              -0.464761   
1                              -0.732626   
2                               0.100699   
3                               0.640011   
4                               0.693578   

   ('angle(tBodyGyroJerkMean,gravityMean)',)  ('angle(X,gravityMean)',)  \
0                                  -0.018446                  -0.841247   
1                                   0.703511                  -0.844788   
2                                   0.808529                  -0.848933   
3                                  -0.485366                  -0.848649   
4                                  -0.615971                  -0.847865   

   ('angle(Y,gravityMean)',)  ('angle(Z,gravityMean)',)  
0                   0.179941                  -0.058627  
1                   0.180289                  -0.054317  
2                   0.180637                  -0.049118  
3                   0.181935                  -0.047663  
4                   0.185151                  -0.043892  

[5 rows x 561 columns]

In [48]:
# Carregar os dados de treino e teste
caminho_teste_x = 'X_test.csv'
dados_teste_x  = pd.read_csv(caminho_teste_x)
dados_teste_x.head()

('tBodyAcc-mean()-X',)  ('tBodyAcc-mean()-Y',)  ('tBodyAcc-mean()-Z',)  \
0                0.257178               -0.023285               -0.014654   
1                0.286027               -0.013163               -0.119083   
2                0.275485               -0.026050               -0.118152   
3                0.270298               -0.032614               -0.117520   
4                0.274833               -0.027848               -0.129527   

   ('tBodyAcc-std()-X',)  ('tBodyAcc-std()-Y',)  ('tBodyAcc-std()-Z',)  \
0              -0.938404              -0.920091              -0.667683   
1              -0.975415              -0.967458              -0.944958   
2              -0.993819              -0.969926              -0.962748   
3              -0.994743              -0.973268              -0.967091   
4              -0.993852              -0.967445              -0.978295   

   ('tBodyAcc-mad()-X',)  ('tBodyAcc-mad()-Y',)  ('tBodyAcc-mad()-Z',)  \
0              -0.952501              -0.925249              -0.674302   
1              -0.986799              -0.968401              -0.945823   
2              -0.994403              -0.970735              -0.963483   
3              -0.995274              -0.974471              -0.968897   
4              -0.994111              -0.965953              -0.977346   

   ('tBodyAcc-max()-X',)  ...  ('fBodyBodyGyroJerkMag-meanFreq()',)  \
0              -0.894088  ...                              0.071645   
1              -0.894088  ...                             -0.401189   
2              -0.939260  ...                              0.062891   
3              -0.938610  ...                              0.116695   
4              -0.938610  ...                             -0.121711   

   ('fBodyBodyGyroJerkMag-skewness()',)  ('fBodyBodyGyroJerkMag-kurtosis()',)  \
0                             -0.330370                             -0.705974   
1                             -0.121845                             -0.594944   
2                             -0.190422                             -0.640736   
3                             -0.344418                             -0.736124   
4                             -0.534685                             -0.846595   

   ('angle(tBodyAccMean,gravity)',)  \
0                          0.006462   
1                         -0.083495   
2                         -0.034956   
3                         -0.017067   
4                         -0.002223   

   ('angle(tBodyAccJerkMean),gravityMean)',)  \
0                                   0.162920   
1                                   0.017500   
2                                   0.202302   
3                                   0.154438   
4                                  -0.040046   

   ('angle(tBodyGyroMean,gravityMean)',)  \
0                              -0.825886   
1                              -0.434375   
2                               0.064103   
3                               0.340134   
4                               0.736715   

   ('angle(tBodyGyroJerkMean,gravityMean)',)  ('angle(X,gravityMean)',)  \
0                                   0.271151                  -0.720009   
1                                   0.920593                  -0.698091   
2                                   0.145068                  -0.702771   
3                                   0.296407                  -0.698954   
4                                  -0.118545                  -0.692245   

   ('angle(Y,gravityMean)',)  ('angle(Z,gravityMean)',)  
0                   0.276801                  -0.057978  
1                   0.281343                  -0.083898  
2                   0.280083                  -0.079346  
3                   0.284114                  -0.077108  
4                   0.290722                  -0.073857  

[5 rows x 561 columns]

In [49]:
# Carregar os rótulos de treino e teste
caminho_treino_y = 'y_train.txt'
dados_treino_y = pd.read_csv(caminho_treino_y, header=None, names=["label"])
dados_treino_y.head()

label
0      5
1      5
2      5
3      5
4      5

In [50]:
# Carregar os rótulos de treino e teste
caminho_teste_y = 'y_test.txt'
dados_teste_y = pd.read_csv(caminho_teste_y, header=None, names=["label"])
dados_teste_y.head()

label
0      5
1      5
2      5
3      5
4      5

### 2. Divisão da base em Treino, Validação e Teste

A base já se encontra dividida em Treino e Validação. O que vamos fazer então é extrair uma base de Validação da base de Treino.

Extraia 25% da base de treino como base de validação.

In [52]:
# Dividindo a base de treino em subconjuntos de treino e validação
dados_treino_x_final, dados_validacao_x, dados_treino_y_final, dados_validacao_y = train_test_split(
    dados_treino_x, 
    dados_treino_y, 
    test_size=0.25, 
    random_state=42
)

# Verificando o tamanho das novas bases de dados
(dados_treino_x_final.shape, dados_validacao_x.shape, dados_treino_y_final.shape, dados_validacao_y.shape)


((5514, 561), (1838, 561), (5514, 1), (1838, 1))

### 3. Melhores 3 variáveis

Rode uma árvore com profundidade máxima igual a 4 para prever a atividade humana com todas as variáveis.
Observe a importância das variáveis e considere as 3 variáveis com maior importância para os próximos passos.
Dica: utilize o atributo ```clf.feature_importances_``` da árvore treinada.

In [55]:
# Criando o modelo de árvore de decisão com profundidade máxima de 4
modelo_arvore = DecisionTreeClassifier(max_depth=4, random_state=42)

# Treinando o modelo com o conjunto de treino final
modelo_arvore.fit(dados_treino_x_final, dados_treino_y_final)

# Extraindo a importância das variáveis
importancias = modelo_arvore.feature_importances_

# Identificando as 3 variáveis mais importantes
indices_importantes = importancias.argsort()[-3:][::-1]
nomes_variaveis_importantes = dados_treino_x.columns[indices_importantes]
importancias_variaveis_importantes = importancias[indices_importantes]

# Exibindo as 3 variáveis mais importantes
nomes_variaveis_importantes, importancias_variaveis_importantes


(Index(['('tGravityAcc-min()-X',)', '('fBodyAccJerk-bandsEnergy()-1,16',)',
        '('tGravityAcc-mean()-Y',)'],
       dtype='object'),
 array([0.28530319, 0.25409067, 0.16460791]))

### 4. Construa uma árvore com as 3 melhores variáveis

Utilizando as três variáveis encontradas acima, construa uma árvore de decisão. Encontre o melhor ```ccp_alpha``` utilizando a base de validação, conforme a estrutura que vimos em aula.

In [58]:
%%time
# Selecionando apenas as 3 variáveis mais importantes para treino e validação
X_treino_reduzido = dados_treino_x_final[nomes_variaveis_importantes]
X_validacao_reduzido = dados_validacao_x[nomes_variaveis_importantes]

# Treinando o modelo de árvore de decisão sem poda
modelo_arvore_reduzido = DecisionTreeClassifier(random_state=42)
modelo_arvore_reduzido.fit(X_treino_reduzido, dados_treino_y_final)

# Calculando a complexidade e o erro de classificação para a base de validação
path = modelo_arvore_reduzido.cost_complexity_pruning_path(X_treino_reduzido, dados_treino_y_final)
ccp_alphas, impurities = path.ccp_alphas, path.impurities

# Testando diferentes valores de ccp_alpha e armazenando a acurácia para cada um
acc_scores = []
for ccp_alpha in ccp_alphas:
    modelo = DecisionTreeClassifier(random_state=42, ccp_alpha=ccp_alpha)
    modelo.fit(X_treino_reduzido, dados_treino_y_final)
    y_pred = modelo.predict(X_validacao_reduzido)
    accuracy = accuracy_score(dados_validacao_y, y_pred)
    acc_scores.append(accuracy)

# Encontrando o ccp_alpha com a melhor acurácia
melhor_alpha = ccp_alphas[acc_scores.index(max(acc_scores))]
melhor_alpha, max(acc_scores)

CPU times: total: 5 s
Wall time: 5.64 s


(0.0002747826469262813, 0.8944504896626768)

### 5. Avaliação do modelo

Avalie a árvore encontrada no item anterior na base de testes.

In [59]:
# Criando e treinando a árvore de decisão com o melhor ccp_alpha encontrado
modelo_final = DecisionTreeClassifier(random_state=42, ccp_alpha=melhor_alpha)
modelo_final.fit(X_treino_reduzido, dados_treino_y_final)

# Selecionando apenas as 3 variáveis mais importantes da base de teste
X_teste_reduzido = dados_teste_x[nomes_variaveis_importantes]

# Fazendo previsões na base de teste
y_pred_teste = modelo_final.predict(X_teste_reduzido)

# Calculando a acurácia na base de teste
acuracia_teste = accuracy_score(dados_teste_y, y_pred_teste)
acuracia_teste

0.6884967763827621

A acurácia da árvore de decisão na base de testes é de aproximadamente 68.85%